addons/binding/org.openhab.binding.unifi/src/main/java/org/openhab/binding/unifi/handler/UniFiControllerHandler.java

```diff
@@ -0,0 +1,172 @@
+            updateStatus(OFFLINE, ThingStatusDetail.CONFIGURATION_ERROR,
+                    e.getMessage() + " - Please double-check your configuration and try again.");
+            controller = null;
+            return;
+        }
+
+        scheduleRefreshJob();
+    }
+
+    @Override
+    public void dispose() {
+        cancelRefreshJob();
+    }
+
+    private void scheduleRefreshJob() {
+        logger.debug("Scheduling refresh job every {}s", config.getRefresh().longValue());
+        refreshJob = scheduler.scheduleWithFixedDelay(this, 0, config.getRefresh().longValue(), TimeUnit.SECONDS);
+    }
+
+    private void cancelRefreshJob() {
+        if (refreshJob != null) {
+            logger.debug("Cancelling refresh job");
+            refreshJob.cancel(true);
+            refreshJob = null;
+        }
+    }
+
+    private void refresh() {
+        ThingStatus status = getThing().getStatus();
+        ThingStatusDetail statusDetail = getThing().getStatusInfo().getStatusDetail();
+        String statusDescription = getThing().getStatusInfo().getDescription();
+
+        if (controller == null) {
+            // mgb: this should never happen
+            logger.warn("UniFi Controller is NULL - this should never happen!");
+        } else {
+            // mgb: we're offline, we need to try to login
+            if (status == INITIALIZING || (status == OFFLINE && statusDetail == COMMUNICATION_ERROR)) {
+                logger.debug("Trying to establish communication with the UniFi controller");
+                if (controller.login()) {
+                    status = ONLINE;
+                    statusDetail = ThingStatusDetail.NONE;
+                    statusDescription = null;
+                }
+            }
+
+            // mgb: only refresh if we're online
+            if (status == ONLINE) {
+                logger.debug("Refreshing the UniFi Controller {}", getThing().getUID());
+
+                try {
+                    // mgb: refresh the controller...
+                    controller.refresh();
+
+                    // mgb: ... then refresh all the clients
+                    for (Thing thing : getThing().getThings()) {
+                        ThingHandler handler = thing.getHandler();
+                        if (handler instanceof UniFiClientHandler) {
+                            UniFiClientHandler clientHandler = (UniFiClientHandler) handler;
+                            clientHandler.refresh();
+                        }
+                    }
+                } catch (UniFiException e) {
+                    logger.warn("Error refreshing the UniFi Controller {} - {}", getThing().getUID(), e.getMessage());
+                    status = OFFLINE;
+                    statusDetail = COMMUNICATION_ERROR;
+                    statusDescription = e.getMessage();
+                }
+            }
+        }
+
+        // mgb: update the status if it's changed
+        if (status != getThing().getStatus()) {
```

Comment:
We might have a different OFFLINE cause now. why should we check?

Reply:
I added the check because without the check, the log gets filled with `ThingStatusInfoEvent` messages:

```
2018-04-12 21:28:48.487 [INFO ] [ome.event.ThingStatusInfoEvent:47   ] - 'unifi:controller:home' updated: ONLINE
2018-04-12 21:28:58.538 [INFO ] [ome.event.ThingStatusInfoEvent:47   ] - 'unifi:controller:home' updated: ONLINE
2018-04-12 21:29:08.596 [INFO ] [ome.event.ThingStatusInfoEvent:47   ] - 'unifi:controller:home' updated: ONLINE
2018-04-12 21:29:18.658 [INFO ] [ome.event.ThingStatusInfoEvent:47   ] - 'unifi:controller:home' updated: ONLINE
```

Reply:
@htreu is this expected I thought we did not log repeated events?

Reply:
Of course that is welcome, note that it will get you in ESH (Eclipse smart home) project